nomad/structs/network.go

```diff
@@ -742,6 +756,12 @@ func getDynamicPortsStochastic(nodeUsed Bitmap, minDynamicPort, maxDynamicPort i
 				goto PICK
 			}
 		}
+		// the pick conflicted with a previous pick that hasn't been saved to
+		// the index yet
+		if portsInOffer != nil && slices.Contains(portsInOffer, randPort) {
```

Comment:
The nil check (and len check above) aren't technically necessary as Contains uses `range` which just doesn't loop over empty slices.

A map or set would work for portsInOffer too, but I suspect it should be small enough that the slice is more/as efficient for common cases (<10 ports per offer).

Reply:
The real advantage of `Set` is it makes code like this more readable, e.g.

`slices.Contains(portsInOffer, randPort)`

becomes

`portsInOffer.Contains(randPort)`

while _also_ being more efficient `O(1)` vs `O(n)` lookups - though it would be interesting to benchmark here!

Reply:
Fixed the nil check (not sure which len check you mean?). I've kept the slice over the map/set because as you noted it doesn't make a meaningful performance difference at the size we're talking about and on top of that when I do the optimization pass mentioned in https://github.com/hashicorp/nomad/pull/16401/files#r1130056010 the return value ends up being a slice anyways.

Reply:
Eh, I was writing this comment still when @shoenig posted. I'm sold... I'll swap it out for a set.